File: src/generator.ts
import axios from "axios";
import fs from "node:fs";
import * as prettier from "prettier";

// This code generates types for all AI modules based on https://api.turing.sh/list
(async () => {
  let res = await axios.get("https://api.turing.sh/list");
  let data = res.data.types;
  let modules = Object.keys(data);
  let models = [];
  modules.forEach((module) => {
    let modelsFromType = data[module];
    modelsFromType.forEach((model) => {
      models.push({
        ...model,
        module: module,
      });
    });
  });
  await generateFiles(modules, models);
})();

async function generateFiles(modules: any[], modelsList) {
  if (!fs.existsSync("./src/modules")) fs.mkdirSync("./src/modules");
  modules.forEach(async (module) => {
    let models = modelsList.map((model) => {
      if (model.module == module) {
        return model;
      } else {
        return {};
      }
    });
    models = models.filter((model) => model.name);
    await writeFileSync(
      `./src/modules/${module}.ts`,
      `import Base from "../base.js";
      import { EventEmitter } from "events";

      export default class ${
        module.charAt(0).toUpperCase() + module.slice(1)
      } extends Base {
        constructor(start: {
            apiKey: string;
            captchaKey: string;
            options?: {
              stream?: boolean;
              host?: string;
            };
        }) {
            super(start);
        }
        ${models
          .map((model) => {
            let nameFn = model.name.split("-")[0];
            let params = Object.keys(model.parameters);
            let type = params
              .map((param) => {
                let parameter = model.parameters[param];
                let t = parameter.type;
                if (t == "array") t = "any[]";
                return `${param}${parameter.required ? "" : "?"}: ${t};`;
              })
              .join("\n");
            let responseType = "any";
            if (model.response) {
              let responseParams = Object.keys(model.response);
              responseType = responseParams
                .map((param) => {
                  let parameter = model.response[param];
                  let t = parameter.type;
                  if (t == "array") t = "any[]";
                  return `${param}${parameter.required ? "" : "?"}: ${t};`;
                })
                .join("\n");
              responseType = `{${responseType}}`;
            }
            let host = "${this.options.host}";
            return `async ${nameFn}(data: {
                ${type}
            }): Promise<EventEmitter | ${responseType}>{
             return await this.fetch(\`${host}/${module}/${model.name}\`, data);
           }`;
          })
          .join("\n")}
      }`
    );
  });
  await writeFileSync(
    `./src/index.ts`,
    `${modules
      .map(
        (module) =>
          `import ${
            module.charAt(0).toUpperCase() + module.slice(1)
          } from "./modules/${module}.js";`
      )
      .join("\n")}

  export {
    ${modules.map((module) => module.charAt(0).toUpperCase() + module.slice(1))}
  }`
  );
}

async function writeFileSync(path, text) {
  fs.writeFileSync(path, await prettier.format(text, { parser: "typescript" }));
}

File: generator.py
import requests
import os

def generate_files(modules, models_list):
    if not os.path.exists("./src/modules"):
        os.makedirs("./src/modules")

    for module in modules:
        models = [model for model in models_list if model['module'] == module]
        models = [model for model in models if model['name']]

        with open(f"./src/modules/{module}.py", "w") as file:
            file.write(f"import Base\n")
            file.write(f"from events import EventEmitter\n")
            file.write(f"\nclass {module.capitalize()}(Base):\n")
            file.write(f"    def __init__(self, start):\n")
            file.write(f"        super().__init__(start)\n")
            for model in models:
                name_fn = model['name'].split("-")[0]
                params = model['parameters']
                response_params = model['response']
                file.write(f"\n    async def {name_fn}(self, data):\n")
                file.write(f"        response = await self.fetch('{module}/{model['name']}', data)\n")
                file.write(f"        return response\n")

def main():
    response = requests.get("https://api.turing.sh/list")
    data = response.json()
    modules = list(data['types'].keys())
    models_list = []
    for module, models_from_type in data['types'].items():
        for model in models_from_type:
            models_list.append({
                **model,
                'module': module
            })

    generate_files(modules, models_list)

    with open("./src/index.py", "w") as index_file:
        index_file.write("\n".join([
            f"from modules.{module} import {module.capitalize()}"
            for module in modules
        ]))
        index_file.write(f"\n\n__all__ = [{', '.join(modules)}]")

if __name__ == "__main__":
    main()
